src/server/game/Movement/MovementGenerators/IdleMovementGenerator.cpp

```diff
@@ -70,74 +114,133 @@ bool RotateMovementGenerator::Update(Unit* owner, uint32 diff)
+void RotateMovementGenerator::Deactivate(Unit*)
 {
-    return ROTATE_MOTION_TYPE;
+    AddFlag(MOVEMENTGENERATOR_FLAG_DEACTIVATED);
 }
 
-void RotateMovementGenerator::Finalize(Unit* owner)
+void RotateMovementGenerator::Finalize(Unit* owner, bool/* active*/, bool movementInform)
 {
-    owner->ClearUnitState(UNIT_STATE_ROTATING);
-    if (owner->GetTypeId() == TYPEID_UNIT)
-        owner->ToCreature()->AI()->MovementInform(ROTATE_MOTION_TYPE, 0);
+    AddFlag(MOVEMENTGENERATOR_FLAG_FINALIZED);
+
+    if (movementInform && owner->GetTypeId() == TYPEID_UNIT)
+        owner->ToCreature()->AI()->MovementInform(ROTATE_MOTION_TYPE, _id);
+}
+
+MovementGeneratorType RotateMovementGenerator::GetMovementGeneratorType() const
+{
+    return ROTATE_MOTION_TYPE;
 }
 
 //----------------------------------------------------//
 
+DistractMovementGenerator::DistractMovementGenerator(uint32 timer, float orientation) : _timer(timer), _orientation(orientation)
+{
+    Mode = MOTION_MODE_DEFAULT;
+    Priority = MOTION_PRIORITY_HIGHEST;
+    Flags = MOVEMENTGENERATOR_FLAG_INITIALIZATION_PENDING;
+    BaseUnitState = UNIT_STATE_DISTRACTED;
+}
+
 void DistractMovementGenerator::Initialize(Unit* owner)
 {
+    RemoveFlag(MOVEMENTGENERATOR_FLAG_INITIALIZATION_PENDING);
+    AddFlag(MOVEMENTGENERATOR_FLAG_INITIALIZED);
+
     // Distracted creatures stand up if not standing
     if (!owner->IsStandState())
         owner->SetStandState(UNIT_STAND_STATE_STAND);
 
-    owner->AddUnitState(UNIT_STATE_DISTRACTED);
+    Movement::MoveSplineInit init(owner);
+    init.MoveTo(PositionToVector3(*owner), false);
+    if (owner->HasUnitMovementFlag(MOVEMENTFLAG_ONTRANSPORT) && owner->GetTransGUID())
+        init.DisableTransportPathTransformations();
+    init.SetFacing(_orientation);
+    init.Launch();
 }
 
-void DistractMovementGenerator::Finalize(Unit* owner)
+void DistractMovementGenerator::Reset(Unit* owner)
 {
-    owner->ClearUnitState(UNIT_STATE_DISTRACTED);
+    RemoveFlag(MOVEMENTGENERATOR_FLAG_DEACTIVATED);
 
-    // If this is a creature, then return orientation to original position (for idle movement creatures)
-    if (owner->GetTypeId() == TYPEID_UNIT && owner->ToCreature())
-    {
-        float angle = owner->ToCreature()->GetHomePosition().GetOrientation();
-        owner->SetFacingTo(angle);
-    }
+    Initialize(owner);
 }
 
-bool DistractMovementGenerator::Update(Unit* /*owner*/, uint32 diff)
+bool DistractMovementGenerator::Update(Unit* owner, uint32 diff)
 {
+    if (!owner)
+        return false;
+
     if (diff > _timer)
+    {
+        AddFlag(MOVEMENTGENERATOR_FLAG_INFORM_ENABLED);
         return false;
+    }
 
     _timer -= diff;
     return true;
 }
 
+void DistractMovementGenerator::Deactivate(Unit*)
+{
+    AddFlag(MOVEMENTGENERATOR_FLAG_DEACTIVATED);
+}
+
+void DistractMovementGenerator::Finalize(Unit* /*owner*/, bool/* active*/, bool/* movementInform*/)
+{
+    AddFlag(MOVEMENTGENERATOR_FLAG_FINALIZED);
+
+    /*
+     *  TODO: ugh
```

Comment:
ugh

Reply:
aka will change with time and future work, something more descriptive would be better though xd